### ISSUE TITLE
"order aware" for-tpl

### ISSUE THREAD

**Issue by thedavidmeister** [hoplon/hoplon#230]
i keep running into this same problem in different ways and am not exactly sure what to do about it in a general sense but thought i'd open an issue to keep track

we currently have `for-tpl` as the standard workhorse for any kind of list. the reason we use this and not a normal `for` or `map` is to cache DOM elements as a simple work around for nasty GC edge cases in JS.

in a `for-tpl`, every change to the list is treated as a brand new list and the DOM is update accordingly, there's an internal cache for the elements as an implementation detail but if you ignore that then every data change is a brand new DOM.

this means that any re-arrangement of elements is an illusion only - under the hood the DOM elements remain _in situ_ and hoplon updates their attributes to match the current state of the list.

to be clear, there's nothing conceptual in `for-tpl` that can differentiate between:

```clojure
(def my-list (j/cell [:a :b]))
(for-tpl [x my-list] ...)
(swap! my-list reverse) ; <-- same elements in different order
```

and

```clojure
(def my-list (j/cell [:a :b]))
(for-tpl [x my-list] ...)
(reset! my-list [:c :d]) ; <-- totally different elements
```

we already have subtle problems caused by the combination of this approach and javelin's propagation cache - documented at https://github.com/hoplon/hoplon/issues/194 and https://github.com/hoplon/javelin/issues/37

but this issue is about the problem of wanting dynamic components downstream of the `for-tpl` to have state that tied to the element regardless of its position in the list rather than the data associated with the cell at the specific position in the list.

for example, imagine i have an element with a simple boolean cell to toggle whether a list item shows an expanded or collapsed view, called `expand?`.

```clojure
(defn my-item
 [item]
 (let [expand? (j/cell false)]
  ...))

(for-tpl [x my-list] (my-item x))
```

now, if i have a list 3 items long, `[:a :b :c]` and i expand `:c` then reverse the list to be `[:c :b :a]` then the DOM will show `:a` expanded, not `:c` as `expand?` is _positional_ only!

this is almost certainly not the desired behaviour.

a workaround that i've used a bit but don't love is to wrap a little state around the item definition and turn `expand?` into a lens:

```clojure
(let [state (j/cell {})]
 (defn my-item
  [item]
  (let [k (j/cell= ...) ; <-- get an id from item
          expand? (j/cell= (get state k) (partial swap! state assoc id))] ; <-- e.g. javelin readme
   ...)))

(for-tpl [x my-list] (my-item x))
```

this will then tie `expand?` to a key/id pulled from the item rather than the position from for-tpl but this has a few drawbacks:

- `my-item` must know ahead of time that it is going to be used in a for-tpl, making it less general
- building the lenses and extra scope feels like "bad boilerplate"

the other approach i've tried is to push this state up, outside the `for-tpl` in the parent and then pass it into the element but this has similar issues, but just moved to the parent's internals instead

as i said above, i don't know the general solution for this, but maybe something along the lines of:

```clojure
(defn my-item
 [item & {:keys [expand?]}]
 (let [expand? (or expand? (j/cell false))]
  ...))

(keyed-for-tpl [[k x] keyed-list]
 (keyed-let [expand? (j/cell false)] ; <-- magically this is a lens on k and an internal state map
  (my-item x :expand? expand?)))

**Comment by thedavidmeister** [comment]
i'm going to make a start on this by trying to abstract the "key + state + lens" approach into a utility fn

**Comment by micha** [comment]
I think these may be relevant:
* https://github.com/candera/vmt/blob/master/src/weathergen/cljs/macros.clj#L169
* https://github.com/candera/vmt/blob/master/src/weathergen/ui/common.cljs.hl#L883

**Comment by thedavidmeister** [comment]
@micha haha, indeed! glad i posted this issue to uncover some prior art

what do we think about moving something like this into core?

**Comment by micha** [comment]
sounds awesome to me, could maybe add kwargs to the `for-tpl` macro or something?

```
(defmacro for-tpl
  [[bindings items & {:keys [keyfn]}] body]
  ...
```

**Comment by thedavidmeister** [comment]
yeah that makes sense to me, i think we should get the mechanics right first (start by copying and pasting the linked example 😄 ) and whip up some basic tests, then review the syntax before we merge it in

**Comment by micha** [comment]
totally

**Comment by thedavidmeister** [comment]
@micha do you know if this definitely works/ed?

i pulled it into core as-is in https://github.com/hoplon/hoplon/pull/231/files and get:

```
WARNING: Can't take value of macro hoplon.core/formula-of at line 670 src/hoplon/core.cljs
WARNING: Can't take value of macro hoplon.core/formula-of at line 694 src/hoplon/core.cljs
WARNING: Can't take value of macro hoplon.core/formula-of at line 694 src/hoplon/core.cljs
```

**Comment by thedavidmeister** [comment]
nm, i think it just wanted me to add the ns as well

**Comment by thedavidmeister** [comment]
so i pulled `keyed-for-tpl` into core in the linked branch ^^

having a bit of a play around locally using `build-jar` to get a feel for it beyond the tests.

so far i'd say that the approach of caching/positioning DOM elements and associating data by _key_ is far more intuitive than moving the data out from underneath elements then caching them by list index/length.

for context i'm trying to build a drag and drop UI with HTML5 dnd + hoplon, and draggable items can be expanded/collapsed as described above.

it promises to fix or alleviate some really gnarly bugs that i wasn't looking forward to tackling such as:

- browser focus state remaining at the position i dragged from rather than where i drag to
- tracking the expand state per item (as in OP and tests)
- weird things that i haven't debugged yet with WYSIWYG instances in the draggable items
- other per-item state issues such as tracking which item is being deleted (there's a countdown timer after the user clicks)
- how to handle an element being rearranged if a new order is synced in realtime via e.g. Ably
- the potential to completely side-step the above mentioned cell graph cache issues

considering it's a WIP on a mostly drop-in replacement and pretty much "just works" i'm excited about the potential 😄 

i will say it isn't quite perfect yet though, some outstanding things to work through:

- chase up meaning of ominous and vague comments like "check with micha" and "issues with hoplon macros"
- more tests
- should items really go to `nil` when the DOM element is detached?
- should/can DOM elements in the cache not in use be eventually deleted/cleaned up in a GC friendly way?
- remove coupling to `goog.dom`
- clean out debug code and improve comments/docs
- i did manage to generate jank in the UI as items re-arranged themselves (hopefully some well placed RAF and/or dosync can help here)
- update syntax and merge with `for-tpl` as per micha's comments (i'm a little worried about BC/upgrade path but let's discuss this)
- i immediately found myself reaching for an even more global/generalised DOM/data cache/binding

this last point needs some explanation.

as i'm creating a drag and drop, i have multiple lists (imagine something kind of like trello or waffle) where i'm dragging items between lists. in context of drag and drop, this use-case is very common and will be presented in the standard demos for most dnd libraries.

dragging an item around within a single list works _much_ better with the `keyed-for-tpl` than a regular `for-tpl` but as soon as i drag an item into a new list the situation becomes bad - immediately i get 2 different copies of the item in the respective list scopes, with their own internal state. Dragging an expanded item around in the first list will keep it expanded, dragging it into a new list will collapse it (default state of the newly created item) then dragging it back to the first list expands it again (restoring the cached original element to the DOM).

i feel like an additional parameter might be warranted - a scope key - that would work a lot like a namespace. without a scope key the tpl would function as it currently does, locally scoped and keyed by item. with a scope key, the items would be keyed globally across that scope so that there would only ever be one DOM element for a given key across all lists with the same scope - it would be the responsibility of the app to ensure that each item key is only present in one list at a time or weird things would start happening as separate tpls fight over a single DOM el (detect and throw an error somehow?) but this hopefully this should be relatively easy to ensure with a sensible setup of cell logic

**Comment by thedavidmeister** [comment]
for context on perf of DOM manipulation of reversing a list - https://github.com/riot/riot/issues/484

**Comment by thedavidmeister** [comment]
removed goog.dom by implementing `IFn` for `DocumentFragment` (it works the same way as `Element`)

**Comment by thedavidmeister** [comment]
i managed to remove a lot of the original code (~2/3rds of it) by lifting the requirement that items are set to `nil` when detached, so it would be good to know why this was important

**Comment by thedavidmeister** [comment]
playing around with this, the way the document fragment blows away the whole list is not ideal as it causes elements to lose focus etc. - not sure if addressing this with a diffing algo is in scope for MVP though

**Comment by thedavidmeister** [comment]
individual items in an unscoped tpl will naturally go to `nil` as soon as that item leaves `items` as that value is no longer stored anywhere (even though the el associated with that item key still exists)

scoped tpls will retain the last value of `item` in the el even if it is removed

**Comment by thedavidmeister** [comment]
@micha I've got something that seems to be working ok, can we push merging it with the existing `for-tpl` to a follow-up issue?

my concern is just that the internals of the two loops are very different and the new one has very little testing in the context of real codebases. I feel like we should road test it a bit more before trying to include it directly in existing fns.

**Comment by thedavidmeister** [comment]
more reading https://reactjs.org/docs/lists-and-keys.html

cliffnotes:
- react has the same state issues if you don't provide a key
- their compiler even throws a warning if you don't provide a key
- if you use a key they have an internal diffing algorithm that can re-arrange elements more efficiently than a keyless approach (sounds cool but i didn't do this bit yet!)
